JujuWindowsUtils/JujuWindowsUtils.psm1

```diff
@@ -352,18 +352,38 @@ function Install-WindowsFeatures {
     )
     PROCESS {
         $rebootNeeded = $false
-        foreach ($feature in $Features) {
-            $state = Start-ExecuteWithRetry -Command {
-                Install-WindowsFeature -Name $feature -ErrorAction Stop
-            }
-            if ($state.Success -eq $true) {
-                if ($state.RestartNeeded -eq 'Yes') {
-                    $rebootNeeded = $true
+            foreach ($feature in $Features) {
+                if (Get-IsNanoServer) {
+                    try {
+                        $featureStat = Start-ExecuteWithRetry -Command {
+                            Get-WindowsOptionalFeature -Online -FeatureName $feature
+                        }
+                    } catch {
+                        Throw "Failed to get '$feature' feature status: $_"
+                    }
+                    if ($featureStat.State -ne "Enabled") {
+                        $featureInstall = Start-ExecuteWithRetry -Command {
+                            Enable-WindowsOptionalFeature -Online -FeatureName $feature -All -NoRestart
+                        }
+                        if ($featureInstall.RestartNeeded) {
+                            $rebootNeeded = $true
+                        }
+                    } elseif ($featureStat.RestartNeeded) {
+                        $rebootNeeded = $true
+                    }
+                } else {
+                    $state = Start-ExecuteWithRetry -Command {
```

Comment:
Give it a shot without the retry. The original code that did the retry was not any wiser either. I think we are better off wrapping this function in a retry if needed.